SPEC.md

```diff
@@ -0,0 +1,334 @@
+# Nexus RPC HTTP Specification
+
+## Overview
+
+The Nexus protocol, as specified below, is a synchronous RPC protocol. Arbitrary length operations are modelled on top
+of a set of pre-defined synchronous RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. The caller can cancel an asynchronous operation, check for its outcome, or fetch its current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation when it
+is ready.
+
+## Operation Addressability
+
+An operation is addressed using three components:
+
+- [Service Name](#service-name)
+- [Operation Name](#operation-name)
+- [Operation ID](#operation-id)
+
+> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.
+
+### Service Name
+
+Service names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation Name
+
+Operation names:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+### Operation ID
+
+Operation IDs:
+
+- MUST not be empty.
+- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].
+
+Valid characters include:
+
+```
+ALPHA / DIGIT / "-" / "." / "_" / "~"
+```
+
+## Schema Definitions
+
+### Failure
+
+The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON
```

Comment:
Are "failed" and "canceled" precisely defined in this context? If not then perhaps double quotes or no quotes rather than backticks here.